File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/PhotoPagerAdapter.java
package de.bahnhoefe.deutschlands.bahnhofsfotos;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.github.chrisbanes.photoview.PhotoView;

import java.util.ArrayList;
import java.util.List;

import de.bahnhoefe.deutschlands.bahnhofsfotos.model.PageablePhoto;

public class PhotoPagerAdapter extends RecyclerView.Adapter<PhotoPagerAdapter.PhotoViewHolder> {

    private final List<PageablePhoto> pageablePhotos = new ArrayList<>();

    private final Context context;

    public PhotoPagerAdapter(Context context) {
        this.context = context;
    }

    public int addPageablePhoto(PageablePhoto pageablePhoto) {
        pageablePhotos.add(pageablePhoto);
        notifyDataSetChanged();
        return pageablePhotos.size() - 1;
    }

    @NonNull
    @Override
    public PhotoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        var view = LayoutInflater.from(context).inflate(R.layout.photo_view_item, parent, false);
        return new PhotoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PhotoViewHolder holder, int position) {
        var pageablePhoto = getPageablePhotoAtPosition(position);
        if (pageablePhoto == null) {
            holder.photoView.setImageResource(R.drawable.photo_missing);
        } else {
            holder.photoView.setImageBitmap(pageablePhoto.getBitmap());
        }
    }

    public PageablePhoto getPageablePhotoAtPosition(int position) {
        return pageablePhotos.isEmpty() ? null : pageablePhotos.get(position);
    }

    @Override
    public int getItemCount() {
        return pageablePhotos.isEmpty() ? 1 : pageablePhotos.size();
    }

    public static class PhotoViewHolder extends RecyclerView.ViewHolder {

        PhotoView photoView;

        public PhotoViewHolder(@NonNull View itemView) {
            super(itemView);
            photoView = itemView.findViewById(R.id.photoView);
        }
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/PhotoPagerAdapter.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.github.chrisbanes.photoview.PhotoView
import de.bahnhoefe.deutschlands.bahnhofsfotos.PhotoPagerAdapter.PhotoViewHolder
import de.bahnhoefe.deutschlands.bahnhofsfotos.model.PageablePhoto

class PhotoPagerAdapter(private val context: Context) : RecyclerView.Adapter<PhotoViewHolder>() {
    private val pageablePhotos: MutableList<PageablePhoto> = mutableListOf()

    @SuppressLint("NotifyDataSetChanged")
    fun addPageablePhoto(pageablePhoto: PageablePhoto): Int {
        pageablePhotos.add(pageablePhoto)
        notifyDataSetChanged()
        return pageablePhotos.size - 1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotoViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.photo_view_item, parent, false)
        return PhotoViewHolder(view)
    }

    override fun onBindViewHolder(holder: PhotoViewHolder, position: Int) {
        val pageablePhoto = getPageablePhotoAtPosition(position)
        if (pageablePhoto == null) {
            holder.photoView.setImageResource(R.drawable.photo_missing)
        } else {
            holder.photoView.setImageBitmap(pageablePhoto.bitmap)
        }
    }

    fun getPageablePhotoAtPosition(position: Int): PageablePhoto? {
        return if (pageablePhotos.isEmpty()) null else pageablePhotos[position]
    }

    override fun getItemCount(): Int {
        return if (pageablePhotos.isEmpty()) 1 else pageablePhotos.size
    }

    class PhotoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var photoView: PhotoView = itemView.findViewById(R.id.photoView)
    }
}